File: susupay-frontend/src/types/viral.ts
export interface ReferralStats {
  referral_code: string;
  total_referrals: number;
  referral_names: string[];
}

export interface AchievementItem {
  achievement_type: string;
  title: string;
  description: string;
  icon: string;
  earned_at: string | null;
  earned: boolean;
}

export interface AchievementListResponse {
  earned: AchievementItem[];
  available: AchievementItem[];
  total_earned: number;
  total_available: number;
}

export interface SavingsGoal {
  id: string;
  title: string;
  target_amount: string;
  current_amount: string;
  progress_percent: number;
  target_date: string | null;
  is_active: boolean;
  created_at: string;
}

export interface SavingsGoalCreate {
  title: string;
  target_amount: number;
  target_date?: string;
}

export interface LeaderboardEntry {
  rank: number;
  client_id: string;
  full_name: string;
  streak: number;
  total_deposits: string;
  is_current_user: boolean;
}

export interface LeaderboardResponse {
  period_label: string;
  entries: LeaderboardEntry[];
  my_rank: number | null;
}

export interface ShareLinkResponse {
  invite_url: string;
  whatsapp_url: string;
  message: string;
}

File: susupay-backend/app/schemas/viral.py
import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, Field


# --- Referral Schemas ---

class ReferralStats(BaseModel):
    referral_code: str
    total_referrals: int
    referral_names: list[str]


# --- Achievement Schemas ---

class AchievementDefinition(BaseModel):
    type: str
    title: str
    description: str
    icon: str


class AchievementItem(BaseModel):
    achievement_type: str
    title: str
    description: str
    icon: str
    earned_at: datetime | None = None
    earned: bool = False


class AchievementListResponse(BaseModel):
    earned: list[AchievementItem]
    available: list[AchievementItem]
    total_earned: int
    total_available: int


# --- Savings Goal Schemas ---

class SavingsGoalCreate(BaseModel):
    title: str = Field(min_length=1, max_length=120)
    target_amount: Decimal = Field(gt=0)
    target_date: date | None = None


class SavingsGoalResponse(BaseModel):
    id: uuid.UUID
    title: str
    target_amount: Decimal
    current_amount: Decimal = Decimal("0.00")
    progress_percent: float = 0.0
    target_date: date | None = None
    is_active: bool
    created_at: datetime

    model_config = {"from_attributes": True}


# --- Leaderboard Schemas ---

class LeaderboardEntry(BaseModel):
    rank: int
    client_id: uuid.UUID
    full_name: str
    streak: int
    total_deposits: Decimal
    is_current_user: bool = False


class LeaderboardResponse(BaseModel):
    period_label: str
    entries: list[LeaderboardEntry]
    my_rank: int | None = None


# --- Share Link Schema ---

class ShareLinkResponse(BaseModel):
    invite_url: str
    whatsapp_url: str
    message: str
